65108: Support JMeter variables in GraphQL HTTP Request

woonsan opened the pull request:
## Description

In GraphQL query/mutation variables, JMeter variables can be referenced.
A JSON String value with a JMeter variable reference (e.g, `{ "var1":"${jm.var1}"}`) has no problem as they are replaced at runtime without breaking any JSON spec, but a non-String value with JMeter variable reference (e.g, `{ "nvar1":${jm.nvar1}}`) causes a problem because it's not a valid JSON. We need to somehow allow to use JMeter variable references in non-String JSON values.

This PR provides a conceptually simple solution:
- If there's any JMeter variable references in a non-String JSON value, then replace the JMeter variable reference with a JSON Reference-like expression temporarily and parse it into a JSON Object.
- Before serializing the GraphQL variables JSON object to a string, restore the temporary JSON Reference-like expression to the original JMeter variable references.

That is,
- Step 1: User input for GraphQL variables:
```
{
  "var1": "hello",
  "var2": ${user.count}
}
```
- Step 2: Temporarily replace it with JSON Reference-like expressions:
```
{
  "var1": "hello",
  "var2": {"$ref":"jmeter#/nonStringVariables/user.count"}
}
```
- Step 3: Parse the string into JSONObject
- Step 4: Before serializing the JSONObject or its parent JSONObject, restore the JMeter variable references from the serialized json string:
```
{
  "var1": "hello",
  "var2": ${user.count}
}
```

## Motivation and Context

See Bugzilla Id: 65108 .
Also see PR #651 .

## How Has This Been Tested?
- Related unit test is updated.
- You can also test the demo JMeter script at xdocs/demos/SimpleGraphQLTestPlan.jmx:
```
# Run ApolloGraphQL starwars-server, an open-source demo Apollo GraphQL server.
./gradlew runGui
# open the JMeter script
# Start the test and verify the test result.
```

## Types of changes
- Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code follows the [code style][style-guide] of this project.
- [ ] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

vlsi commented:
Just wondering: does this escape-unescape buy us much?
Would it be better to keep the string as is, and perform parsing/conversion at the runtime only?

woonsan commented:
@vlsi , "performing parsing/conversion at runtime only" seems very difficult as @FSchumacher mentioned "they would have needed more changes, than I would liked to have" in PR #651 . So I ended up this "conceptually simple" solution instead.

FSchumacher commented:
I am not sure, that this is a better solution to not doing any JSON parsing at all.

The problems arise (in my eyes) from the fact, that we re-used a sampler, that wasn't thought for it and can't be used without change. One idea to modify it, was to add a listener/lifecycle-handler interface to that specific sampler, where we could register a callback, that would get called at the time the sampling gets done. There we could parse the JSON if needed. But that would mean adding new interfaces and changes to the public interface of the sampler.

The idea of checking the JSON at edit time, is probably doomed, as JMeter is too flexible. A user could try to use the following code in the JSON value field: `${json_prefix} "abc": 42 ${json_suffix}` with `json_prefix={` and `json_suffix=}`. That would not be handled with the proposed patch.

woonsan commented:
Thanks for the remarks! I fully understand the pointers and there must be better designs and solutions than this quick solution, indeed.
Just two things though:
- The solution includes re-formatting to remove unnecessary white spaces like most GraphQL clients does, and this is done by parsing and formatting through the Jackson library. This "re-formatting" can possibly be done in a different way, but I assume the current solution is simpler.
- The prefix and suffix in the solution are not just simple prefix/suffix, but instead it's following the JSON Reference syntax pattern, which means it is very unlikely that GraphQL variable JSON data could possibly allow any JSON Reference at all, or even allowed in the future, it is very unlikely for users to refer to the "JMeter-specific" json object reference in GraphQL variables in reality.
- Of course, JMeter is too flexible, so for that reason, I actually narrow down the scope only to numeric variable value expression in this solution, which I believe should be good enough practically.

Regards,

Woonsan

FSchumacher commented:
So should we include this fix?

FSchumacher commented on src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/GraphQLRequestParamUtils.java:
```diff
@@ -56,6 +55,12 @@
     private static final Pattern WHITESPACES_PATTERN = Pattern.compile("\\p{Space}+");
 
+    private static final Pattern JMETER_VARIABLE_REFERENCE_PATTERN = Pattern
+            .compile("(:\\s*)\\$\\{([^\\$\\{\\}]+)\\}([\\s,\\}])");
```
The part `[^\\$\\{\\}]+` seems not correct to me. I believe, that you wanted to simulate a *negative lookahead* here.
You probably want to match a complete variable reference like `${varname}` and not two or more like in `${var1} ${var2}`.
What about more complex setups like `${function(${var1})}`?

woonsan replied:
Hi @FSchumacher ,

I'm very sorry for late response.
My intent there is to find a string, (a) starting with "${", followed by (b) any char(s) excluding '$', '{' and '}', followed by (c) "}".  The negative part is only for (b): var name shouldn't include the special chars. I believe it should work even if it looks a bit complex.

If we want to create Pattern objects every time instead of having just one generic Pattern, I'm not so sure how it could improve the readability because the varname should probably be dynamically concatenated at runtime to build a regex pattern, which might be less readable in the end, IMHO.

Regarding the complex setup example with function, is it possible to use that complex variable expression in JMeter? (If so, perhaps we might need to replace the regex by commons-text's StrSubstitutor instead.

Regards,

Woonsan

FSchumacher replied:
No problem :)

About the regex. In regex a `[` starts a character group, which means, that the order of the elements inside the brackets is irrelevant. 
If you look at the following examples:
```
abc: ${__log("Hello\, ${__machineIP()}")}
abc: ${__log("Some $ and some { or even }")}
abc: ${someFunction(abc)}${anotherFunction}
abc: ${func2(a,b)}
abc: ${func1(a)}
abc: ${func0()}
abc: ${more(abc,)}
abc: Nothing to be found here
```
 `(:\s*)\$\{([^\$\{\}]+)\}([\s,\}])` would match the part after `abc` in lines
```
abc: ${func2(a,b)}
abc: ${func1(a)}
abc: ${func0()}
abc: ${more(abc,)}
```
and a regex with negative lookahead like `:\s*\$\{(?!.*\$\{).*\)}` would match 
```
abc: ${__log("Some $ and some { or even }")}
abc: ${func2(a,b)}
abc: ${func1(a)}
abc: ${func0()}
abc: ${more(abc,)}
```
And those examples don't include a simple variable like `${var}`, which would yield wrong results with the lookahead variant.

About the complex function setup. Considering the usage of the `__log` function, I think it is not that complex. The following expression is a valid one: `${__log("Hello\, ${__machineIP()}")}`.
Another example is given in the help for [`__eval`](https://jmeter.apache.org/usermanual/functions.html#__eval)

Given this, I don't think, that a simple regex approach would help all cases. It might be enough to help most users, though.

woonsan replied:
I see. Thanks a lot for the explanations, @FSchumacher !
Perhaps shall we just avoid Jackson object serialization/deserialization, without having to validate or reformat the json variables at all? It would make it work at least...

FSchumacher replied:
I would favour dropping the step, as I can't see a good way to keep the functionality *and* make it work in all cases.

woonsan replied:
Hi @FSchumacher ,

I've updated my PR, not to parse/serialize variables, and now it seems working fine with any var expressions in variables or even in query.

Please have a look.

Thanks in advance,

Woonsan

vincebrown commented:
Hey JMeter team / @woonsan ! 
wondering if there is going to be any movement on this pr , or if you have any suggestions on temporary work arounds for this issue in the mean time? 

Thank you and the team for all the work you put in.

vlsi commented on src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/GraphQLRequestParamUtils.java:
```diff
@@ -79,28 +80,24 @@ public static boolean isGraphQLContentType(final String contentType) {
         }
 
-        postBodyJson.set(QUERY_FIELD, JsonNodeFactory.instance.textNode(StringUtils.trim(params.getQuery())));
+        appendJsonStringField(sb, QUERY_FIELD, escapeNewLinesForJson(StringUtils.trim(params.getQuery())));
```
Why only newlines are escaped? What if the value contains quotes, slashes, or something else?

woonsan replied:
Pushed a new commit: a524a04fa69b40870e346ff805df51b7e5a4ba22 in order to use `JsonStringEncoder` to escape json string, which seems working fine for string including quotes, backslashes, new lines, etc.

vlsi replied:
What do you think of using `JsonGenerator` to generate the json?

woonsan replied:
Fixed using `JsonGenerator` for safety: 5c9fb4f92b20bb3feb5770023d7ed4ed021a29d6 Thanks!

vlsi commented on src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/GraphQLRequestParamUtils.java:
```diff
@@ -273,4 +264,13 @@ private static String getJsonNodeTextContent(final JsonNode jsonNode, final bool
         throw new IllegalArgumentException("Not a string value node.");
     }
+
+    private static void appendJsonStringField(final StringBuilder sb, final String fieldName, final String fieldValue) {
```
Thi seems no longer used

woonsan replied:
Removed. Thanks!

vlsi commented on src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/GraphQLRequestParamUtils.java:
```diff
@@ -79,28 +81,27 @@ public static boolean isGraphQLContentType(final String contentType) {
-            } catch (JsonProcessingException e) {
-                log.error("Ignoring the GraphQL query variables content due to the syntax error: {}",
-                        e.getLocalizedMessage());
+        final JsonFactory jsonFactory = new JsonFactory();
```
https://fasterxml.github.io/jackson-core/javadoc/2.6/com/fasterxml/jackson/core/JsonFactory.html

Please move the factory to a static field

woonsan replied:
Done in 9e63931ee11f206fa846bd8277bfe8b8bd6cb13d. Thanks!

vlsi commented on src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/GraphQLRequestParamUtils.java:
```diff
@@ -79,28 +81,27 @@ public static boolean isGraphQLContentType(final String contentType) {
-                log.error("Ignoring the GraphQL query variables content due to the syntax error: {}",
-                        e.getLocalizedMessage());
+        final JsonFactory jsonFactory = new JsonFactory();
+        final StringWriter writer = new StringWriter(1024);
```
The default looks odd.
I would go with dropping the default, or you could have better estimate based on the actual length of `params.getVariables()`, `params.getQuery()`, etc, etc.

The problem with hard-coding constants is that no one knows the reasons for 1024 to be used here, and no one would know when to revise the constant :-/

woonsan replied:
I see. Just removed the default size in 9e63931ee11f206fa846bd8277bfe8b8bd6cb13d.

vlsi commented on src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/GraphQLRequestParamUtils.java:
```diff
@@ -79,28 +81,27 @@ public static boolean isGraphQLContentType(final String contentType) {
+            gen.writeStringField(OPERATION_NAME_FIELD, StringUtils.trimToNull(params.getOperationName()));
+
+            if (StringUtils.isNotBlank(params.getVariables())) {
+                gen.writeRaw(",\"" + VARIABLES_FIELD + "\":");
```
Just wondering: does writeFieldName work here?

woonsan replied:
I tried with that at first to find out it didn't work. When writing a raw data, fieldName should be written together.

vlsi replied:
This works:

```java
                gen.writeFieldName(VARIABLES_FIELD);
                gen.writeRawValue(StringUtils.trim(params.getVariables()));
```

vlsi commented on src/protocol/http/src/test/java/org/apache/jmeter/protocol/http/util/TestGraphQLRequestParamUtils.java:
```diff
@@ -99,7 +103,15 @@ void testInvalidGraphQLContentType(String contentType) {
+        final JsonNode expectedPostBodyJson = objectMapper.readTree(EXPECTED_POST_BODY);
+        final JsonNode actualPostBodyJson = objectMapper.readTree(GraphQLRequestParamUtils.toPostBodyString(params));
+
+        assertEquals(expectedPostBodyJson.get(GraphQLRequestParamUtils.OPERATION_NAME_FIELD),
+                actualPostBodyJson.get(GraphQLRequestParamUtils.OPERATION_NAME_FIELD));
```
Could you please factor this into a assertion method that receives `fieldName`, `expectedPostBody`, and `params`?

Then assertion code would be de-duplicated, and the assertion message could include the source value, so the failure includes not only "expected A got B", but it should include which field failed the comparison, and what was the input data.

vlsi replied:
An alternative option is to use `@ParameterizedTest`, however, please include `field name` and `input data` into the assertion message in any case

woonsan replied:
Hi @vlsi ,
I tried to improve it using `@ParamterizedTest` in 9e63931ee11f206fa846bd8277bfe8b8bd6cb13d. Could you take a look? Thanks!

vlsi commented on src/protocol/http/src/test/java/org/apache/jmeter/protocol/http/util/TestGraphQLRequestParamUtils.java:
```diff
@@ -97,9 +102,24 @@ void testInvalidGraphQLContentType(String contentType) {
+        final JsonNode expectedPostBodyJson = objectMapper.readTree(EXPECTED_POST_BODY);
+        final JsonNode actualPostBodyJson = objectMapper.readTree(
+                GraphQLRequestParamUtils.toPostBodyString(new GraphQLRequestParams(OPERATION_NAME, QUERY, VARIABLES)));
+        return Stream.of(
+                Triple.of(GraphQLRequestParamUtils.OPERATION_NAME_FIELD, expectedPostBodyJson, actualPostBodyJson),
+                Triple.of(GraphQLRequestParamUtils.VARIABLES_FIELD, expectedPostBodyJson, actualPostBodyJson),
+                Triple.of(GraphQLRequestParamUtils.QUERY_FIELD, expectedPostBodyJson, actualPostBodyJson));
```
Please use `Arguments.arguments` as suggested in JUnit5 documentation: https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests-sources-MethodSource

woonsan replied:
Ah, that's brilliant!

vlsi commented on src/protocol/http/src/test/java/org/apache/jmeter/protocol/http/util/TestGraphQLRequestParamUtils.java:
```diff
@@ -97,9 +102,24 @@ void testInvalidGraphQLContentType(String contentType) {
+
+    @ParameterizedTest
+    @MethodSource("postBodyFieldNameAndJsonNodes")
+    void testToPostBodyString(Triple<String, JsonNode, JsonNode> comparisonTriple) throws Exception {
+        final String fieldName = comparisonTriple.getLeft();
+        final JsonNode expectedNode = comparisonTriple.getMiddle();
+        final JsonNode actualNode = comparisonTriple.getRight();
```
Please use three different arguments instead of `Triple`

woonsan replied:
Done: ff3eba435347a3148332aa22a0a6d6d0e08646ff. Thanks!

vlsi commented on src/protocol/http/src/main/java/org/apache/jmeter/protocol/http/util/GraphQLRequestParamUtils.java:
```diff
@@ -46,16 +48,18 @@
  */
 public final class GraphQLRequestParamUtils {
 
-    private static final String VARIABLES_FIELD = "variables";
+    static final String VARIABLES_FIELD = "variables";
```
Should this be public then?

woonsan replied:
Makes sense. Done: ff3eba435347a3148332aa22a0a6d6d0e08646ff

vlsi commented on src/protocol/http/src/test/java/org/apache/jmeter/protocol/http/util/TestGraphQLRequestParamUtils.java:
```diff
@@ -97,9 +102,21 @@ void testInvalidGraphQLContentType(String contentType) {
-    @Test
-    void testToPostBodyString() throws Exception {
-        assertEquals(EXPECTED_POST_BODY, GraphQLRequestParamUtils.toPostBodyString(params));
+    static Stream<org.junit.jupiter.params.provider.Arguments> postBodyFieldNameAndJsonNodes() throws Exception {
+        final JsonNode expectedPostBodyJson = objectMapper.readTree(EXPECTED_POST_BODY);
+        final JsonNode actualPostBodyJson = objectMapper.readTree(
+                GraphQLRequestParamUtils.toPostBodyString(new GraphQLRequestParams(OPERATION_NAME, QUERY, VARIABLES)));
+        return Stream.of(
+                arguments(GraphQLRequestParamUtils.OPERATION_NAME_FIELD, expectedPostBodyJson, actualPostBodyJson),
+                arguments(GraphQLRequestParamUtils.VARIABLES_FIELD, expectedPostBodyJson, actualPostBodyJson),
+                arguments(GraphQLRequestParamUtils.QUERY_FIELD, expectedPostBodyJson, actualPostBodyJson));
+    }
+
+    @ParameterizedTest
+    @MethodSource("postBodyFieldNameAndJsonNodes")
+    void testToPostBodyString(String fieldName, JsonNode expectedNode, JsonNode actualNode) throws Exception {
+        assertEquals(expectedNode, actualNode,
+                "The value of the '" + fieldName + "' field doesn't match in " + actualNode);
```
Please refine the test so it is clear what it does.

Currently, it looks like it duplicates the same data 3 times, and `fieldName` seems to be unused.
What is the purpose of having parameters here?

woonsan replied:
It is for testing whether the `#toPostBodyString()` has the equivalent JSON data by asserting each field, `operationName`, `variables` and `query`. And the `fieldName` is used in the assert message to explain which field is possibly different from the expectation.

woonsan replied:
Thanks! I've improved it a bit : a00a07f6aeca4a247625b9d4212ce35adb49a5a7

Commit:
65108: Support JMeter variables in GraphQL HTTP Request

Previously the variables were parsed as JSON, and now the variable
contents is added as a raw part (no parsing) of the JSON.

fixes #660

haideralii commented:
@woonsan can you describe how can I use the jmete variable in graphQL variables?
like I have below variable how can I use groupId as integer, same like topic Id which is currently static
{
"group_ids": ["${groupId}"]
"topic_ids": [1560],
}

woonsan commented:
@haideralii , this was originally to fix the problem reported at https://bz.apache.org/bugzilla/show_bug.cgi?id=65108. Please see the example in the ticket. And if needed, please continue discussion in either JMeter user list or the bugzilla. Thanks.
